File: rust/src/problems/p543.rs
use crate::utils::tree::TreeNode;
use std::cell::RefCell;
use std::rc::Rc;

pub struct Solution;

impl Solution {
    pub fn diameter_of_binary_tree(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        fn dfs(root: &Option<Rc<RefCell<TreeNode>>>) -> (i32, i32) {
            match root {
                Some(n) => {
                    let n = n.borrow_mut();
                    let (depth_left, diameter_left) = dfs(&n.left);
                    let (depth_right, diameter_right) = dfs(&n.right);
                    let depth = depth_left.max(depth_right) + 1;
                    let diameter = diameter_left
                        .max(diameter_right)
                        .max(depth_left + depth_right);
                    (depth, diameter)
                }
                None => (0, 0),
            }
        }
        dfs(&root).1
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn case1() {
        let root = TreeNode::from_vec(&[1, 2, 3, 4, 5]);
        let result = Solution::diameter_of_binary_tree(root);
        assert_eq!(3, result);
    }
}

File: cpp/p543.cpp
#include "helpers.h"
#include <cassert>
#include <vector>

using namespace std;

class Solution {
    int ans;
    int dfs(TreeNode *root) {
        if (!root) {
            return 0;
        }
        int l = dfs(root->left);
        int r = dfs(root->right);
        ans = max(ans, l + r + 1);
        return max(l, r) + 1;
    }

public:
    int diameterOfBinaryTree(TreeNode *root) {
        ans = 0;
        dfs(root);
        return ans - 1;
    }
};

int main() {
    vector<int> nodes = {1, 2, 3, 4, 5};
    auto root = parse_tree(nodes);
    int r = Solution().diameterOfBinaryTree(root);
    assert(r == 3);
    return 0;
}
